### ISSUE TITLE
job-exec: raise job exception for jobs on ranks that go offline

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#3906]
Once a mechanism is in place to take ranks offline that are unresponsive (#3897), the job-exec module should take advantage of this and raise a job exception for any jobs which were running on that rank. This can be a fatal job exception at first, but in the future we may want to allow any other shells in the job to decide if they want to keep running.

**Comment by grondo** [comment]
Something that @garlick brought up in a discussion in #3448:

> a lost broker could conceivably rejoin and recover a running job, once we support that

Specifically, we want to support restarting a broker without killing local jobs. In this case the job-exec module should _not_ raise an exception on the job. Determining the difference between a down rank which may rejoin and recover the running job, and a rank which is unlikely to do may be a challenge.

I suggested in that thread

> Initially, all top-level jobs in the system instance will themselves be instances of Flux, so maybe we leave it up to the job to terminate itself when its own brokers go offline? This solves the case of restarting system brokers when system nodes are still online, as the sub-instance shouldn't lose connectivity.

Though I'm not sure if that is a great idea because misbehaving instance could tie up resources if it doesn't exit.

An alternative might be to have the exec system wait "a long time" after a rank or ranks go offline before raising a fatal job exception. This would cover the case where a broker is restarted quickly after a crash or manual restart.

**Comment by garlick** [comment]
Short term, it is probably straightforward to have `job-exec` watch the `online` broker group like `resource` now does, and have it raise fatal exceptions on jobs running on nodes that transition to offline.

**Comment by grondo** [comment]
>  and have it raise fatal exceptions on jobs running on nodes that transition to offline.

Maybe there should also be a configuration flag to disable this behavior? Otherwise, as discussed above, it will be difficult to support broker restart without losing running jobs (#3801)  if running jobs are killed when the restarted broker goes offline.

**Comment by garlick** [comment]
Good point.  Right now PMI will limit our ability to restart a broker in a non-system instance, so this could be on by default and disabled in the system instance config.

**Comment by grondo** [comment]
During recent testing, I found that if a broker is disconnected from the overlay, this results in an immediate `EHOSTUNREACH` error for each remote shell running on the affected subtree, which in turn causes the job-exec module to raise an exception. So (depending on the meaning of the word `offline` here), perhaps this issue is already solved?

Example:
```console
$ src/cmd/flux start -o,-Stbon.fanout=4 -s 4 --test-exit-mode=leader sh -x -c 'id=$(flux mini submit -N4 -n4 sleep inf) && flux job wait-event $id start && flux python t/scripts/startctl.py kill 3 9 && flux job attach -vE $id'
+ flux mini submit -N4 -n4 sleep inf
+ id=ƒ3H2M3Y3
+ flux job wait-event ƒ3H2M3Y3 start
1637015172.881775 start
+ flux python t/scripts/startctl.py kill 3 9
flux-start: 3 (pid 1709) Killed
+ flux job attach -vE ƒ3H2M3Y3
0.000s: job.submit {"userid":1000,"urgency":16,"flags":0}
0.013s: job.depend
0.013s: job.priority {"priority":16}
0.015s: job.alloc {"annotations":{"sched":{"resource_summary":"rank[0-3]/core0"}}}
0.096s: job.start
2021-11-15T22:26:14.620709Z job-exec.err[0]: remote_exec_cb: flux_rpc_get_unpack: rank 3: No route to host
1.836s: job.exception type=exec severity=0 cmd=/home/grondo/git/f.git/src/shell/flux-shell: rank=3 exec failed: No route to host
flux-job: task(s) exited with exit code 143
6.842s: job.finish {"status":36608}
6.845s: job.release {"ranks":"all","final":true}
6.846s: job.free
6.846s: job.clean
```

The `flux_rexec()` RPC gets this error immediately since it acts as a streaming RPC, and the stream is terminated once the child broker becomes disconnected. So, this "accidentally" started working when @garlick added support for RPC tracking for TBON children (8cc6bee52022d8eac298f4e5f32d8fc34bcef111?)

**Comment by garlick** [comment]
That makes sense - it was good foresight to handle the failing RPC by raising the exception!

**Comment by garlick** [comment]
BTW  maybe at least for the EHOSTUNREACH case, that error message could be made more human friendly, e.g.  "lost contact with job shell on broker rank 3, job will die" or similar.

**Comment by grondo** [comment]
Good idea (I assume you mean the job exception error message). Maybe I can tack that improvement onto #3962 (and rename the PR to something like `job-exec: better handling of remote exec errors`)

**Comment by garlick** [comment]
I meant this one, for log spelunkers:
> 2021-11-15T22:26:14.620709Z job-exec.err[0]: remote_exec_cb: flux_rpc_get_unpack: rank 3: No route to host

My thought was just that calling out the internal functions doesn't really improve clarity, and "No route to host" might put someone on the wrong track (IP routing problem?)

**Comment by grondo** [comment]
That error is logged from libsubprocess so I don't think we can call out whether we lost contact with the "job shell" vs some other process that is being executed at the time.

Also, this is just one of many, many places where internal functions are output in log error messages (almost 600 matches for `git grep __FUNCTION__`). Perhaps we want a new issue to clean that up globally at some point.

**Comment by garlick** [comment]
Ooof!  Ok, sorry, I hadn't looked that closely.  Never mind.

**Comment by grondo** [comment]
OTOH, users will probably be _very_ confused when `No route to host` pops up in an exception that killed their job, and at the level from which the job exception is issued we _do_ know we "lost" a job shell, so we can actually issue a better error message.

I wonder if job exceptions of this nature should also be logged, in case the logs are all that we have available when doing post mortem?

**Comment by garlick** [comment]
Good thought!  Yeah, that would be useful.

Maybe libsubprocess shouldn't be directly logging _likely_ errors, and instead leave that up to the callers, if that's possible here.

(Admittedly, what was once unlikely suddenly became likely after we started generating EHOSTUNREACH errors.)

**Comment by garlick** [comment]
Maybe `EHOSTDOWN 112 Host is down` would have been a better choice for the overlay network errors. :shrug:

**Comment by grondo** [comment]
Yeah, or can't error responses actually contain _textual_ error messages instead of numeric only? We could actually put the exact error in these responses rather than trying to fit Flux errors into system errno error message space.

**Comment by garlick** [comment]
Oh, duh.  Looks like I did set "overlay disconnect" as the textual error for the responses generated to upstream RPCs but not for the downstream RPCs.

**Comment by grondo** [comment]
One other thought that occurs to me is that we may want a way to easily determine if a job was killed due to a broker disconnect (i.e. node failure). Should we actually issue a different exception `type` for jobs killed from EHOSTUNREACH? I'm not sure if something like `node-fail` would necessarily be appropriate, but right now all exceptions raised by the exec system have `type=exec`.

**Comment by grondo** [comment]
I think @garlick was right that libsubprocess shouldn't log errors to the flux dmesg log. I'll open a separate issue on fixing that.

Meanwhile, I think this specific issue is already handled so I'll close it (#3962 adds a test)

We do need to track a configuration option to control job exceptions for system instance jobs in the new execution system, so I'll get an issue open on that as well.
